exist-core/src/main/java/org/exist/xquery/functions/securitymanager/IdFunction.java

```diff
@@ -77,14 +77,20 @@ private org.exist.dom.memtree.DocumentImpl functionId() {
 
             builder.startElement(new QName("id", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
 
-            builder.startElement(new QName("real", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
-            subjectToXml(builder, context.getRealUser());
-            builder.endElement();
+            final Subject realUser = context.getRealUser();
+            if (realUser != null) {
+              builder.startElement(new QName("real", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
+              subjectToXml(builder, realUser);
+              builder.endElement();
+            }
 
-            if (!sameUserWithSameGroups(context.getRealUser(), context.getEffectiveUser())) {
-                builder.startElement(new QName("effective", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
-                subjectToXml(builder, context.getEffectiveUser());
-                builder.endElement();
+            final Subject effectiveUser = context.getEffectiveUser();
+            if (effectiveUser != null) {
```

Comment:
I would prefer to combine the nested conditions to

```
if (effectiveUser != null && (
  realUser == null || !sameUserWithSameGroups(realUser, effectiveUser)
)) {
```

Reply:
Additionally the method `sameUserWithSameGroups` could handle null values for both realUser and effectiveUser

Reply:
I thought about that, but to me the conditions mean different things, and I kept them separate.
Either way, the result, and maybe the bytecode, is exactly the same.

I also thought about checking for `null` inside `sameUserWithSameGroups`, which might make the intention of the code clearer.

Reply:
If both are null then it should return `false`.

Reply:
This would simplify the condition to 
```
if (sameUserwithSameGroups(realUser, effectiveUser)) {...}
```

Reply:
:D I'll shut up until you have completed your review.

Reply:
I should have combined them into one instead of brain-dumping :)